custreamz_notebooks/random_haproxy_data_generator.ipynb

```diff
@@ -0,0 +1,194 @@
+    "        'accept_date','frontend_name','backend_name','server_name','time_request',\\\n",
+    "        'time_queue','time_backend_connect', 'time_backend_response', 'time_duration',\\\n",
+    "        'http_status_code', 'bytes_read', 'captured_request', 'captured_response',\\\n",
+    "        'termination_state','actconn','feconn','beconn','srvconn','retries','srv_queue','backend_queue']\n",
+    "\n",
+    "def generate_log():\n",
+    "    log_skelton = \"[haproxy@{0}] <134>{1} {2}[{3}]: {4}:{5} [{6}] {7} {8}/{9} {10}/{11}/{12}/{13}/{14} {15} {16} {17} {18} {19} {20}/{21}/{22}/{23}/{24} {25}/{26}\"\n",
+    "    values = []\n",
+    "    for idx, col in enumerate(cols):\n",
+    "        if col in sample_data:\n",
+    "            value_list = sample_data[col]\n",
+    "            values.append(value_list[randrange(len(value_list))])\n",
+    "        else:\n",
+    "            values.append(values[-1])\n",
+    "    dict_out = {}    \n",
+    "    dict_out[\"logline\"] = log_skelton.format(*values)\n",
+    "    return json.dumps(dict_out)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Below is the example of a logline generated by generate_log()."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "'{\"logline\": \"[haproxy@15.2.6.9] <134>May 28 2019 00:00:51 haproxy[352465]: 26.245.255.1:52465 [28/May/2019:00:02:10.461] px-http px-http/srv3 0/1/2/7/20 200 573 - - PH-- 4/8/1/1/0 0/2\"}'"
+      ]
+     },
+     "execution_count": 3,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "generate_log()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Let's now create a Kafka producer."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "producer_conf = {'bootstrap.servers': 'localhost:9092', \\\n",
+    "                'compression.type':'snappy'}\n",
+    "producer = confluent_kafka.Producer(producer_conf)\n",
+    "topic = \"random-haproxy\"\n",
+    "\n",
+    "def delivery_callback(err, msg):\n",
+    "        if err:\n",
+    "            sys.stderr.write('%% Message failed delivery: %s\\n' % err)\n",
+    "        else:\n",
+    "            sys.stderr.write('%% Message delivered to %s [%d] @ %o\\n' %\n",
+    "                             (msg.topic(), msg.partition(), msg.offset()))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Let's now generate 5k random loglines to Kafka."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "count = 0\n",
+    "try:\n",
+    "    while count < 5000:\n",
+    "        producer.produce(topic, generate_log(), callback=delivery_callback)\n",
```

Comment:
Since you are using an async callback here the script actually exits before the entire 5,000 messages are delivered to Kafka. Easiest thing to do would just be remove the callback since this is just for example purposes and not a lecture on the best Kafka practices.

Reply:
Fixed